In [1]:
import numpy as np

In [2]:
#Correct Running Code for Simpsons 1/3 in Two Dimensions

def g(x, y):
    return np.exp(-x*y) + np.sin(x) - np.cos(y)

def SimpsonsDouble13(g, ax, bx, ay, by, nx, ny):
    hx = (bx - ax) / nx
    hy = (by - ay) / ny
    integration = 0.0

    for i in range(0,nx+1):
        for j in range(0, ny+1):
            x = ax + i * hx
            y = ay + j * hy

            if i==0 or i==nx:
                coefficientx =  1.0
            elif i%2 == 0:
                coefficientx =  2.0
            else:
                coefficientx =  4.0

            if j==0 or j==ny:
                coefficienty =  1.0
            elif j%2 ==0:
                coefficienty =  2.0
            else:
                coefficienty =  4.0

            coefficient = coefficientx * coefficienty
            integration += coefficient*g(x,y)


    integration *= (hx * hy) / 9.0

    return integration

# Define the region of integration [ax, bx] x [ay, by]
ax, bx = 0, 3      # Limits in x
ay, by = 1, 5      # Limits in y
nx, ny = 10,10     # Number of divisions x and y must be divisible by 2

result = SimpsonsDouble13(g,ax,bx,ay,by,nx,ny)
print("The 2 dimensional approximation using Simpson's 1/3 Rule is: ", result)

The 2 dimensional approximation using Simpson's 1/3 Rule is:  14.964915739020096



Lower Limit of x is equal to 0
Upper Limit of x is equal to 3
Lower Limit of y is equal to 1
Upper Limit of y is equal to 5
- nx = ny = 10 the value is 14.964915739020096
- nx = ny = 20 the value is 14.95805414637237
- nx = ny = 40 the value is 14.95757799397997
- nx = ny = 80 the value is  14.957547372024624
- nx = ny = 160 the value is  14.957545444109074

The true value is 14.957545315341

Therefore, the error vector would be
[.0073704237, .000508831031, .000032678638, .000002056683, .000000128768]


In [3]:
ErrorVec = [.0073704237, .000508831031, .000032678638, .000002056683, .000000128768]
def GetConvergence(vec,factor=2):
    for i in range(len(vec)-1):
        print("Convergence = ", (np.log(vec[i]/vec[i+1]))/np.log(factor))
GetConvergence(ErrorVec)

Convergence =  3.8564889962161857
Convergence =  3.96076689555642
Convergence =  3.9899565041175564
Convergence =  3.99747342551032


In [4]:
#Correct Running Code for Simpsons 3/8 in Two Dimensions

def g(x, y):
    return np.exp(-x*y) + np.sin(x) - np.cos(y)

def SimpsonDouble38(g,ax,bx,ay,by,nx,ny):
    hx = (bx - ax) / nx
    hy = (by - ay) / ny
    integration = 0.0

    for i in range(0,nx+1):
        for j in range(0, ny+1):
            x = ax + i * hx
            y = ay + j * hy

            if i==0 or i==nx:
                coefficientx =  1.0
            elif i%3 == 0:
                coefficientx =  2.0
            else:
                coefficientx =  3.0

            if j==0 or j==ny:
                coefficienty =  1.0
            elif j%3 ==0:
                coefficienty =  2.0
            else:
                coefficienty =  3.0

            coefficient = coefficientx * coefficienty
            integration += coefficient*g(x,y)

    integration *= (9 * hx * hy)/64
    return integration

# Define the region of integration [ax, bx] x [ay, by]
ax, bx = 0, 3      # Limits in x
ay, by = 1, 5      # Limits in y
nx, ny = 144,144     # Number of divisions x and y must be divisible by 3

result =  SimpsonDouble38(g,ax,bx,ay,by,nx,ny)
print("The 2 dimensional approximation using Simpson's 3/8 Rule is: ", result)



The 2 dimensional approximation using Simpson's 3/8 Rule is:  14.957545756553609


Lower Limit of x is equal to 0
Upper Limit of x is equal to 3
Lower Limit of y is equal to 1
Upper Limit of y is equal to 5
- nx = ny = 9 the value is 14.97921585919191
- nx = ny = 18 the value is 14.95920560018956
- nx = ny = 36 the value is 14.957655890874488
- nx = ny = 72 the value is  14.957552344398486
- nx = ny = 144 the value is  14.957545756553609

The true value is 14.957545315341

Therefore, the error vector would be
[.0216705439, .0016602848, .000110575533, .000007029057, .000000441212]

In [5]:
ErrorVec2 = [.0216705439, .0016602848, .000110575533, .000007029057, .000000441212]
def GetConvergence(vec,factor=2):
    for i in range(len(vec)-1):
        print("Convergence = ", (np.log(vec[i]/vec[i+1]))/np.log(factor))
GetConvergence(ErrorVec2)

Convergence =  3.7062327202367062
Convergence =  3.9083266366471765
Convergence =  3.9755572320901646
Convergence =  3.9937872193456365
